heron/api/src/scala/Builder.scala

```diff
@@ -0,0 +1,46 @@
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//  Copyright 2017 Twitter. All rights reserved.
+package com.twitter.heron.streamlet.scala
+
+import com.twitter.heron.streamlet.impl.BuilderImpl
+import Builder._
+
+
+object Builder {
+  def newBuilder(): Builder = new BuilderImpl()
+}
+
+/**
+  * Builder is used to register all sources. Builder thus keeps track
+  * of all the starting points of the computation dag and uses this
+  * information to build the topology
+  */
+trait Builder {
+
+  /**
+    * All sources of the computation should register using addSource.
+    * @param supplier The supplier function that is used to create the streamlet
+    */
+  def newSource[R](supplier: SerializableSupplier[R]): Streamlet[R]
```

Comment:
In the light of Scala User perspective and in my opinion, All exposed Scala `Traits` should be **pure scala**. This means Scala traits should accept `Scala Functions` instead of `FunctionalInterface` here:

`def newSource[R](supplier: () => R): Streamlet[R]`

instead of 

`def newSource[R](supplier: SerializableSupplier[R]): Streamlet[R]`